common/src/ipc/transmit.rs

```diff
@@ -288,6 +305,131 @@ impl<T> IpcSocket<T> {
                 .context(IpcErrorKind::MalformedMsg)?;
             Some(Mmap::from_fd(file, len))
         };
-        Ok(RawMsg { code, shm })
+        Ok(IpcMessage { code, shm })
+    }
+}
+
+impl From<Request<'_>> for IpcMessage {
+    fn from(value: Request) -> Self {
+        match value {
+            Request::Ping => Self {
+                code: Code::ReqPing,
+                shm: None,
+            },
+            Request::Query => Self {
+                code: Code::ReqQuery,
+                shm: None,
+            },
+            Request::Kill => Self {
+                code: Code::ReqKill,
+                shm: None,
+            },
+            Request::Clear(clear) => {
+                let mut mmap = Mmap::create(clear.size());
+                clear.serialize(&mut Cursor::new(mmap.slice_mut()));
+                Self {
+                    code: Code::ReqClear,
+                    shm: Some(mmap),
+                }
+            }
+            Request::Img(image) => {
+                let mut mmap = Mmap::create(image.size());
+                image.serialize(&mut Cursor::new(mmap.slice_mut()));
+                Self {
+                    code: Code::ReqImg,
+                    shm: Some(mmap),
+                }
+            }
+        }
+    }
+}
+
+impl<'a> From<&'a IpcMessage> for Request<'a> {
+    fn from(value: &'a IpcMessage) -> Self {
+        match value.code {
+            Code::ReqPing => Self::Ping,
+            Code::ReqQuery => Self::Query,
+            Code::ReqKill => Self::Kill,
+            Code::ReqClear => {
+                let mmap = value.shm.as_ref().expect("clear request must contain data");
+                let clear = ClearRequest::deserialize(&mut Cursor::new(mmap.slice()));
+                Self::Clear(clear)
+            }
+            Code::ReqImg => {
+                let mmap = value.shm.as_ref().expect("image request must contain data");
+                let image = ImageRequest::deserialize(&mut Cursor::new(mmap.slice()));
+                Self::Img(image)
+            }
+            _ => unreachable!("`Request` builder reached invalid state"),
```

Comment:
Same question as below about the `unreachable!` with the invalid states

Reply:
See above. Thank you very much for reviewing, I would wait for owners opinion on additional dependencies and the get to work. If you would want to add some changes to this PR yourself, can you, or do I have to add you somehow?

Reply:
I'd also like to thank you for you input, @itsjunetime. I've had a hard time reviewing this thoroughly, so your help is much appreciated.